gbcamextract.c

```diff
@@ -1,352 +1,245 @@
-      drawSpan( pixelBuffer, lowBits, highBits, x, y );
-    }
-    // second-from-leftmost tile
-    tileNum = frameBuffer[frameAddress + 0x650 + yTile*4 + 1];
-    tileAddress = frameAddress + tileNum*16;
-    // draw the tile
-    for( spanNum = 0; spanNum<8; spanNum++ )
-    {
-      spanAddress = tileAddress + 2*spanNum;
-      lowBits = frameBuffer[spanAddress];
-      highBits = frameBuffer[spanAddress+1];
-      x = 8;
-      y = 16 + yTile*8 + spanNum;
-      drawSpan( pixelBuffer, lowBits, highBits, x, y );
-    }
-    // second-from-rightmost tile
-    tileNum = frameBuffer[frameAddress + 0x650 + yTile*4 + 2];
+    tileNum = frameBuffer[frameAddress + 0x600 + xTile + 0x14*z];
     tileAddress = frameAddress + tileNum*16;
-    // draw the tile
-    for( spanNum = 0; spanNum<8; spanNum++ )
-    {
-      spanAddress = tileAddress + 2*spanNum;
-      lowBits = frameBuffer[spanAddress];
-      highBits = frameBuffer[spanAddress+1];
-      x = 144;
-      y = 16 + yTile*8 + spanNum;
-      drawSpan( pixelBuffer, lowBits, highBits, x, y );
-    }
-    // rightmost tile
-    tileNum = frameBuffer[frameAddress + 0x650 + yTile*4 + 3];
-    tileAddress = frameAddress + tileNum*16;
-    // draw the tile
-    for( spanNum = 0; spanNum<8; spanNum++ )
-    {
-      spanAddress = tileAddress + 2*spanNum;
-      lowBits = frameBuffer[spanAddress];
-      highBits = frameBuffer[spanAddress+1];
-      x = 152;
-      y = 16 + yTile*8 + spanNum;
-      drawSpan( pixelBuffer, lowBits, highBits, x, y );
-    }
+    tile = frameBuffer + tileAddress;
+    x = xTile*8;
+    y = ((z&1)?8:0) + ((z&2)?128:0);
+    drawSpan( pixelBuffer, tile, x, y );
   }
 }
 
-void drawSpan( char pixelBuffer[static WIDTH*HEIGHT], char lowBits, char highBits, int x, int y )
-{
-  int pixelNum;
-  char spanColors[8];
-  int grays[4] = {0xFF, 0xAA, 0x55, 0x00};
-  for(pixelNum=0; pixelNum<8; pixelNum++)
-  {
-    int color = 0;
-    int mask = 1<<pixelNum;
-    if( lowBits & mask )
-      color = 1;
-    if( highBits & mask )
-      color += 2;
-    spanColors[7-pixelNum] = grays[ color ];
-  }
-  for(pixelNum=0; pixelNum<8; pixelNum++)
+void drawSpan( char pixelBuffer[], char *buffer, int x, int y ) {
+  unsigned char lowBits, highBits;
+  char *p, *q;
+  p = pixelBuffer + (x/4) + y * ROW_SIZE;
+  for( q = p + ROW_SIZE * 8; p<q; p+=ROW_SIZE )
   {
-    // copy span to pixel buffer
-    int pixelAddress = x + 160 * y +pixelNum;
-    pixelBuffer[pixelAddress] = spanColors[pixelNum];
+    lowBits = ~*buffer++;
+    highBits = ~*buffer++;
+    p[1] = (lowBits & 1) | ((highBits & 1) << 1) | ((lowBits & 2) << 1) | ((highBits & 2) << 2) | ((lowBits & 4) << 2)  | ((highBits & 4) << 3) | ((lowBits & 8) << 3) | ((highBits & 8) << 4);
+    lowBits >>= 4;
+    highBits >>= 4;
+    p[0] = (lowBits & 1) | ((highBits & 1) << 1) | ((lowBits & 2) << 1) | ((highBits & 2) << 2) | ((lowBits & 4) << 2)  | ((highBits & 4) << 3) | ((lowBits & 8) << 3) | ((highBits & 8) << 4);
```

Comment:
This is the crazy bitwise math that packs eight pixels into two bytes.
I don't know what witchcraft Big N was using when they split the high bits and low bits across two bytes, but it _really_ made this a pain to work with.
If we were going to be converting tens of thousands of images at a time, I am sure a lookup table would be faster. As it is though, this seems to work well enough. My only concern is how difficult it may be to read.